qiskit/transpiler/runningpassmanager.py

```diff
@@ -13,99 +13,124 @@
 """RunningPassManager class for the transpiler.
 This object holds the state of a pass manager during running-time."""
 from __future__ import annotations
-import logging
+
 import inspect
-from functools import partial, wraps
+import logging
+from functools import wraps
 from typing import Callable
 
 from qiskit.circuit import QuantumCircuit
 from qiskit.converters import circuit_to_dag, dag_to_circuit
-from qiskit.dagcircuit import DAGCircuit
-from qiskit.passmanager import BasePassRunner
+from qiskit.passmanager.base_optimization_tasks import OptimizerTask
+from qiskit.passmanager.exceptions import PassManagerError
+from qiskit.utils.deprecation import deprecate_func
+
+# pylint: disable=unused-import
 from qiskit.passmanager.flow_controllers import (
-    PassSequence,
+    BaseFlowController,
     FlowController,
-    DoWhileController,
+    FlowControllerLiner,
+    # for backward compatibility
     ConditionalController,
+    DoWhileController,
 )
-from qiskit.passmanager.exceptions import PassManagerError
-from qiskit.transpiler.basepasses import BasePass
+
 from .exceptions import TranspilerError
-from .fencedobjs import FencedPropertySet, FencedDAGCircuit
 from .layout import TranspileLayout
 
 logger = logging.getLogger(__name__)
 
 
-class RunningPassManager(BasePassRunner):
-    """A RunningPassManager is a running pass manager."""
+class RunningPassManager(FlowControllerLiner):
+    """A RunningPassManager is a running pass manager.
```

Comment:
My understanding of this is that we're pivoting away from "`PassManager` is to `RunningPassManager` as `list` is to `list_iterator`" into "calling `Task.execute` creates and manages iterator state".  I think that's a very sensible API choice, and much more extensible in the future - it also resolves the problematic break where we had `PassManager/RunningPassManager` but no such `FlowController/RunningFlowController` split, which meant the iteration state was not really isolated in the intended way anyway, and required lots of hacking together to inject the pipeline state.  This looks like a much better plan to me.

Reply:
Yeah I think this is what you proposed in the first round of review. I also like this design.